Crash on apply the second time

I have a simple plan that worked before updating to v0.15.0. 

``` shell
> terraform -version                                         
Terraform v0.15.0
on windows_amd64
+ provider registry.terraform.io/hashicorp/helm v2.1.0
+ provider registry.terraform.io/hashicorp/kubernetes v2.0.3
```

The plan deploys to a local kubernetes running in Docker for Windows PostgreSQL and Airflow. There is nothing special about it. Here is a [public gist](https://gist.github.com/mill5james/261e7863a54c05df8a1543a52e32a774) to the code.

This will work the first time, when the environment does not exist.
When running the second time (even when the Terraform plan is unmodified)  I get the following crash:

```
panic: value is marked, so must be unmarked first

goroutine 359 [running]:
github.com/zclconf/go-cty/cty.Value.assertUnmarked(...)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.8.1/cty/marks.go:123
github.com/zclconf/go-cty/cty.Value.LengthInt(0x308a028, 0xc000bf8df0, 0x29342e0, 0xc0006cb0c8, 0x308a001)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.8.1/cty/value_ops.go:1045 +0x56
github.com/zclconf/go-cty/cty.transform(0xc000bf93a0, 0x1, 0x1, 0x308a028, 0xc000bf8df0, 0x29342e0, 0xc0006cb0c8, 0x3053950, 0xc0006cb128, 0xc0006cb0c8, ...)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.8.1/cty/walk.go:144 +0x2f9
github.com/zclconf/go-cty/cty.transform(0x0, 0x0, 0x0, 0x3089ff0, 0xc000bf9300, 0x2845800, 0xc000bed4a0, 0x3053950, 0xc0006cb128, 0x0, ...)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.8.1/cty/walk.go:209 +0x10ed
github.com/zclconf/go-cty/cty.TransformWithTransformer(...)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.8.1/cty/walk.go:125
github.com/zclconf/go-cty/cty.Value.MarkWithPaths(0x3089ff0, 0xc000bf9300, 0x2845800, 0xc000bed4a0, 0xc000d86f80, 0x1, 0x1, 0x0, 0xc000d86f80, 0x1, ...)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.8.1/cty/marks.go:215 +0xcb
github.com/hashicorp/terraform/terraform.markProviderSensitiveAttributes(0xc000a4ba40, 0x3089ff0, 0xc000bf9300, 0x2845800, 0xc000bed4a0, 0x0, 0xc, 0xc00016d0f0, 0xa)
        /home/circleci/project/project/terraform/evaluate.go:960 +0xc5
github.com/hashicorp/terraform/terraform.(*evaluationStateData).GetResource(0xc0018a27e0, 0x4d, 0xc00016d0e0, 0xc, 0xc00016d0f0, 0xa, 0xc00016c8a9, 0x7, 0x88, 0x10, ...)
        /home/circleci/project/project/terraform/evaluate.go:787 +0x1353
github.com/hashicorp/terraform/lang.(*Scope).evalContext(0xc00100f180, 0xc0003ab300, 0x8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x30)
        /home/circleci/project/project/lang/eval.go:360 +0x208d
github.com/hashicorp/terraform/lang.(*Scope).EvalContext(...)
        /home/circleci/project/project/lang/eval.go:238
github.com/hashicorp/terraform/lang.(*Scope).EvalBlock(0xc00100f180, 0x3088ab8, 0xc001797d10, 0xc000a4ba40, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
        /home/circleci/project/project/lang/eval.go:51 +0x105
github.com/hashicorp/terraform/terraform.(*BuiltinEvalContext).EvaluateBlock(0xc00075b930, 0x3088d58, 0xc001797d10, 0xc000a4ba40, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/circleci/project/project/terraform/eval_context_builtin.go:273 +0x1bb
github.com/hashicorp/terraform/terraform.(*NodeAbstractResourceInstance).plan(0xc0018a61c0, 0x30c0990, 0xc00075b930, 0xc0006b3180, 0xc0014a8f60, 0x0, 0x0, 0x0, 0x0, 0xc00016cf77, ...)
        /home/circleci/project/project/terraform/node_resource_abstract_instance.go:577 +0x3df
github.com/hashicorp/terraform/terraform.(*NodeApplyableResourceInstance).managedResourceExecute(0xc000c102a0, 0x30c0990, 0xc00075b930, 0x0, 0x0, 0xc00062dc10)
        /home/circleci/project/project/terraform/node_resource_apply_instance.go:235 +0x63d
github.com/hashicorp/terraform/terraform.(*NodeApplyableResourceInstance).Execute(0xc000c102a0, 0x30c0990, 0xc00075b930, 0xc00009a001, 0xc00062dd18, 0x7ec06a, 0x288f5c0)
        /home/circleci/project/project/terraform/node_resource_apply_instance.go:129 +0xbc
github.com/hashicorp/terraform/terraform.(*ContextGraphWalker).Execute(0xc0008a7380, 0x30c0990, 0xc00075b930, 0x2326381e9b0, 0xc000c102a0, 0x0, 0x0, 0x0)
        /home/circleci/project/project/terraform/graph_walk_context.go:127 +0xc6
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x2bdb460, 0xc000c102a0, 0x0, 0x0, 0x0)
        /home/circleci/project/project/terraform/graph.go:59 +0xbb3
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc00048a540, 0x2bdb460, 0xc000c102a0, 0xc0010089c0)
        /home/circleci/project/project/dag/walk.go:381 +0x2a2
created by github.com/hashicorp/terraform/dag.(*Walker).Update
        /home/circleci/project/project/dag/walk.go:304 +0x1246




!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.471382662.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

SECURITY WARNING: the "crash.471382662.log" file that was created may contain
sensitive information that must be redacted before it is safe to share
on the issue tracker.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
[crash.577544246.log](https://github.com/hashicorp/terraform/files/6313533/crash.577544246.log)

--- a/terraform/context_apply2_test.go
+++ b/terraform/context_apply2_test.go
@@ -8,6 +8,7 @@ import (
 	"time"
 
 	"github.com/hashicorp/terraform/addrs"
+	"github.com/hashicorp/terraform/configs/configschema"
 	"github.com/hashicorp/terraform/providers"
 	"github.com/hashicorp/terraform/states"
 	"github.com/zclconf/go-cty/cty"
@@ -366,4 +367,81 @@ resource "aws_instance" "bin" {
 		t.Fatalf("foo should have no deps after apply, but got %s", foo.Current.Dependencies)
 	}
 
+}
+
+func TestContext2Apply_additionalSensitiveFromState(t *testing.T) {
+	// Ensure we're not trying to double-mark values decoded from state
+	m := testModuleInline(t, map[string]string{
+		"main.tf": `
+variable "secret" {
+  sensitive = true
+  default = ["secret"]
+}
+
+resource "test_resource" "a" {
+  sensitive_attr = var.secret
+}
+
+resource "test_resource" "b" {
+  value = test_resource.a.id
+}
+`,
+	})
+
+	p := new(MockProvider)
+	p.GetProviderSchemaResponse = getProviderSchemaResponseFromProviderSchema(&ProviderSchema{
+		ResourceTypes: map[string]*configschema.Block{
+			"test_resource": {
+				Attributes: map[string]*configschema.Attribute{
+					"id": {
+						Type:     cty.String,
+						Computed: true,
+					},
+					"value": {
+						Type:     cty.String,
+						Optional: true,
+					},
+					"sensitive_attr": {
+						Type:      cty.List(cty.String),
+						Optional:  true,
+						Sensitive: true,
+					},
+				},
+			},
+		},
+	})
+
+	state := states.BuildState(func(s *states.SyncState) {
+		s.SetResourceInstanceCurrent(
+			mustResourceInstanceAddr(`test_resource.a`),
+			&states.ResourceInstanceObjectSrc{
+				AttrsJSON: []byte(`{"id":"a","sensitive_attr":["secret"]}`),
+				AttrSensitivePaths: []cty.PathValueMarks{
+					{
+						Path:  cty.GetAttrPath("sensitive_attr"),
+						Marks: cty.NewValueMarks("sensitive"),
+					},
+				},
+				Status: states.ObjectReady,
+			}, mustProviderConfig(`provider["registry.terraform.io/hashicorp/test"]`),
+		)
+	})
+
+	ctx := testContext2(t, &ContextOpts{
+		Config: m,
+		Providers: map[addrs.Provider]providers.Factory{
+			addrs.NewDefaultProvider("test"): testProviderFuncFixed(p),
+		},
+		State: state,
+	})
+
+	_, diags := ctx.Plan()
+	if diags.HasErrors() {
+		t.Fatal(diags.ErrWithWarnings())
+	}
+
+	_, diags = ctx.Apply()
+	if diags.HasErrors() {
+		t.Fatal(diags.ErrWithWarnings())
+	}
 }

--- a/terraform/evaluate.go
+++ b/terraform/evaluate.go
@@ -782,9 +782,15 @@ func (d *evaluationStateData) GetResource(addr addrs.Resource, rng tfdiags.Sourc
 
 		val := ios.Value
 
-		// If our schema contains sensitive values, mark those as sensitive
+		// If our schema contains sensitive values, mark those as sensitive.
+		// Since decoding the instance object can also apply sensitivity marks,
+		// we must remove and combine those before remarking to avoid a double-
+		// mark error.
 		if schema.ContainsSensitive() {
-			val = markProviderSensitiveAttributes(schema, val)
+			var marks []cty.PathValueMarks
+			val, marks = val.UnmarkDeepWithPaths()
+			marks = append(marks, getValMarks(schema, val, nil)...)
+			val = val.MarkWithPaths(marks)
 		}
 		instances[key] = val
 	}
@@ -952,12 +958,6 @@ func moduleDisplayAddr(addr addrs.ModuleInstance) string {
 	default:
 		return addr.String()
 	}
-}
-
-// markProviderSensitiveAttributes returns an updated value
-// where attributes that are Sensitive are marked
-func markProviderSensitiveAttributes(schema *configschema.Block, val cty.Value) cty.Value {
-	return val.MarkWithPaths(getValMarks(schema, val, nil))
 }
 
 func getValMarks(schema *configschema.Block, val cty.Value, path cty.Path) []cty.PathValueMarks {

--- a/terraform/evaluate_test.go
+++ b/terraform/evaluate_test.go
@@ -563,7 +563,7 @@ func evaluatorForModule(stateSync *states.SyncState, changesSync *plans.ChangesS
 	}
 }
 
-func TestMarkProviderSensitive(t *testing.T) {
+func TestGetValMarks(t *testing.T) {
 	schema := &configschema.Block{
 		Attributes: map[string]*configschema.Attribute{
 			"unsensitive": {
@@ -647,7 +647,7 @@ func TestMarkProviderSensitive(t *testing.T) {
 		},
 	} {
 		t.Run(fmt.Sprintf("%#v", tc.given), func(t *testing.T) {
-			got := markProviderSensitiveAttributes(schema, tc.given)
+			got := tc.given.MarkWithPaths(getValMarks(schema, tc.given, nil))
 			if !got.RawEquals(tc.expect) {
 				t.Fatalf("\nexpected: %#v\ngot:      %#v\n", tc.expect, got)
 			}

